### PR TITLE
Index columns used for order_by

### PR DESCRIPTION
**Pull request by HebaruSan**
## Background

I've had moderate experience with databases in general, but I'm not as well-versed in SQL specifically (make of that what you will), so I'll describe my understanding of this both to help with code review but also so an SQL expert can check the reasoning more easily, should one happen to stop by.

### What is an index?

An index stores a mapping from the values stored in a column to the rows that have each value, separately and in addition to the main storage for the table. Without an index, finding rows with a particular value in a column requires looping over all the rows in the table, loading the value of the column for each row, and checking whether it matches. With an index, you can retrieve the matching rows in one step without checking the other rows.

An index can also speed up sort operations. To sort by a column without an index, you have to retrieve all the values for that column from all the rows, then sort by them. To sort by a column that has an index, you can use the index to find the rows with the largest (or smallest) value, then the rows with the next largest (or smallest) value, and so on, directly and without having to access all rows **or** perform an explicit sorting operation.

References:

- https://www.postgresql.org/docs/current/indexes-ordering.html
- https://use-the-index-luke.com/sql/sorting-grouping/indexed-order-by#:~:text=Indexing%20Order%20By,cover%20the%20order%20by%20clause

## Changes

Now all of the columns that SpaceDock passes to `.order_by()` in various places have `index=True` added to their definitions (`User.username` already has it). This should allow the database to execute those queries more quickly.

- `Featured.created`
- `BlogPost.created`
- `User.created`
- `Game.name`
- `Game.created`
- `Mod.created`
- `Mod.updated`
- `DownloadEvent.created`
- `FollowEvent.created`
- `ReferralEvent.created`

An alembic script is included to handle migration.

(Noticed during review of #277.)

### Measurements

@godarklight tested this on an instance with thousands of test mods (thanks!!), and saw a 3% increase in database size:

```
before update:
 pg_database_size 
------------------
         17015663

After update
 pg_database_size 
------------------
         17638255
```

A copy of prod is more like 28%:

```
postgres=# select pg_database_size('spacedockreal');
 pg_database_size 
------------------
        817251183

postgres=# select pg_database_size('spacedockreal');
 pg_database_size 
------------------
       1054229359
```

Most of the routes got somewhat faster, a few much faster, and even fewer got slightly slower:

| Route | Percentage | dae8e0d | 8f7b22c | Change | 
| --- | --: | --: | --: | --: | 
| _api_browse | 100% | 134.165 | 81.241 | -39% |
| _api_browse_new | 50% | 65.466 | 58.986 | -9% |
| _api_browse_featured | 100% | 3.978 | 4.422 | 11% |
| _api_mod_5 | 100% | 9.258 | 7.404 | -20% |
| _api_mod_5_5 | 100% | 5.152 | 4.413 | -14% |
| _api_user_godarklight | 50% | 5645.066 | 5122.597 | -9% |
| _api_user_godarklight | 100% | 5645.066 | 5122.597 | -9% |
| _ | 50% | 24.081 | 22.492 | -6% |
| _ | 100% | 43.2 | 30.494 | -29% |
| _ksp | 50% | 129.045 | 109.918 | -14% |
| _ksp | 100% | 160.444 | 133.42 | -16% |
| _browse | 100% | 16418.576 | 15648.439 | -4% |
| _browse_new | 50% | 168.856 | 142.925 | -15% |
| _browse_new | 100% | 200.551 | 207.23 | 3% |
| _browse_new.rss | 50% | 12.291 | 6.921 | -43% |
| _browse_new.rss | 100% | 15.79 | 8.107 | -48% |
| _browse_updated | 50% | 164.026 | 146.639 | -10% |
| _browse_updated | 100% | 236.655 | 155.23 | -34% |
| _browse_updated.rss | 50% | 18.621 | 11.842 | -36% |
| _browse_updated.rss | 100% | 45.422 | 30.767 | -32% |
| _browse_top.rss | 50% | 19.242 | 18.068 | -6% |
| _browse_top.rss | 100% | 44.01 | 22.183 | -49% |
| _browse_all | 100% | 18173.835 | 15443.933 | -15% |
| _ksp_browse | 100% | 16075.972 | 14723.338 | -8% |
| _ksp_browse_new | 50% | 178.063 | 144.944 | -18% |
| _ksp_browse_new | 100% | 237.807 | 152.777 | -35% |
| _ksp_browse_new.rss | 50% | 14.339 | 8.482 | -40% |
| _ksp_browse_new.rss | 100% | 18.194 | 10.761 | -40% |
| _ksp_browse_updated | 50% | 169.343 | 148.194 | -12% |
| _ksp_browse_updated | 100% | 217.753 | 156.396 | -28% |
| _ksp_browse_updated.rss | 50% | 20.162 | 13.217 | -34% |
| _ksp_browse_updated.rss | 100% | 24.103 | 14.774 | -38% |
| _ksp_browse_top | 100% | 18879.554 | 14540.464 | -22% |
| _ksp_browse_top.rss | 50% | 18.158 | 16.575 | -8% |
| _ksp_browse_top.rss | 100% | 39.237 | 17.654 | -55% |
| _ksp_browse_featured | 50% | 27.167 | 25.17 | -7% |
| _ksp_browse_featured | 100% | 30.085 | 27.579 | -8% |
| _ksp_browse_featured.rss | 50% | 7.898 | 7.487 | -5% |
| _ksp_browse_featured.rss | 100% | 10.53 | 9.16 | -13% |
| _ksp_browse_all | 100% | 17652.835 | 14326.701 | -18% |

### CODE REVIEW

**Comment by DasSkelett** [comment]
First of all, really good idea!

---

> As far as I can tell, sqlalchemy will take care of populating the index data for us automatically:
> 
> [stackoverflow.com/a/58026283/2422988 :](https://stackoverflow.com/a/58026283/2422988)
> > Now you can just add index=True to an existing column, and auto-generate the migration.

I understand that link differently. I think it means alembic can automatically generate **the revision**, **not** that the database automatically knows which columns to index.

My suspicion seams to be confirmed by the fact that when I run `./generate_revision.sh`, the following db schema upgrade script is generated (excerpt):
```python3
    op.create_index(op.f('ix_blog_created'), 'blog', ['created'], unique=False)
    op.create_index(op.f('ix_downloadevent_created'), 'downloadevent', ['created'], unique=False)
    op.create_index(op.f('ix_featured_created'), 'featured', ['created'], unique=False)
    op.create_index(op.f('ix_followevent_created'), 'followevent', ['created'], unique=False)
    op.create_index(op.f('ix_game_created'), 'game', ['created'], unique=False)
    op.create_index(op.f('ix_game_name'), 'game', ['name'], unique=False)
    op.create_index(op.f('ix_mod_created'), 'mod', ['created'], unique=False)
    op.create_index(op.f('ix_mod_updated'), 'mod', ['updated'], unique=False)
    op.create_index(op.f('ix_referralevent_created'), 'referralevent', ['created'], unique=False)
    op.create_index(op.f('ix_user_created'), 'user', ['created'], unique=False)
```

<details>
<summary>Also looking at Adminer it conveniently reports indexed columns:</summary>

| This is with the new code but without database upgrade: | This is with the new code after an database upgrade with the auto-generated script: |
| --- | --- |
| before | after |

</details>

In short, we need to add a database migration script by running `./generate_revision.sh`.

---

This certainly doesn't come without cost, which I guess is mostly increased memory usage. Since my experience with databases is still basically zero, I don't know how they/Postgres handle memory usage in general, and how much these indices need specifically.
Is there a risk that we could run the db container out of memory with this patch? (We probably have to ask VITAS to get some values for current memory usage).
Should we test this beforehand? Or is it negligible?

**Comment by HebaruSan** [comment]
> In short, we need to add a database migration script by running `./generate_revision.sh`.

Pushed your file in the latest commit (with type hints for the old files). Does it look OK? I'm completely new to alembic, so I'm guessing about pretty much everything.

> This certainly doesn't come without cost, which I guess is mostly increased memory usage. 
> Should we test this beforehand? Or is it negligible?

I think it'd be about the same as adding one new copy of the field to each table.
@godarklight, do you still have that test instance with thousands of automatically-created mods? Maybe we can collect some stats from that.

**Comment by HebaruSan** [comment]
> This certainly doesn't come without cost, which I guess is mostly increased memory usage.
> Should we test this beforehand? Or is it negligible?

The disk space impact is an increase of 3% in the test instance. I believe the db can load and unload parts of the data into and out of RAM as it pleases, so I don't think there's really a definable impact on memory usage. In normal operations, it should be slightly improved as fewer rows need to be loaded from the database, in proportion to the numbers for the speeds of the routes.

**Comment by DasSkelett** [comment]
Nice! I'll do one more test to see if the database updates and downgrades fine, otherwise I'd say this looks good!

**Comment by DasSkelett** [comment]
Pushed a commit that updates the description of the alembic script, which you can see if you run e.g. `alembic history`.

**Comment by HebaruSan** [comment]
image
